rewrite-test/src/main/kotlin/org/openrewrite/java/typegeneration/TypeGenerationTest.kt

```diff
@@ -0,0 +1,512 @@
+import org.openrewrite.InMemoryExecutionContext
+import org.openrewrite.Parser
+import org.openrewrite.java.JavaParser
+import org.openrewrite.java.tree.J
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+import java.io.PrintWriter
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.*
+
+
+interface TypeGenerationTest {
+
+    // This test provides a generic way to assert that a java source (as a string or as a path) can first be
+    // compiled using the runtime classpath and then it will extract the types from the resulting AST and attempt
+    // to compile the source again using just the generated byte code.
+    //
+    // The two variants of assertCompilationWithByteCode only require a source, but allows for the parser, and two
+    // method callbacks to be overridden to assert test conditions.
+    //
+    // The system.out is captured and passed to the compiler out handler (and the default is to swallow that output)
+    // checkCompilationUnit can be use used to verify type attribution on the compilation unit after the second parse.
+
+    val rewriteRoot: Path
+        get () {
+            val normalize = Paths.get(TypeGenerationTest::class.java.getResource("./")!!.toURI())
+                .resolve("../../../../../../../../").normalize()
+            return normalize
+        }
+
+    val parser: JavaParser
+        get() = JavaParser.fromJavaVersion()
+            .logCompilationWarningsAndErrors(true)
+            .build()
+
+    //Default is to swallow compiler output, this can be overridden for any given test.
+    val defaultCompilerOutputHandler: (String) -> Unit
+        get() = {}
+
+    @Test
+    fun testCompilationUsingExtractedTypes() {
+        assertCompilationWithByteCode(
+            source = """
+                    package org.example;
+                    
+                    import org.openrewrite.internal.StringUtils;
+                    import java.util.SortedMap;
+                    import java.util.TreeMap;
+                    
+                    public class Example {
+                        public void foo() {
+                            SortedMap<Integer, Long> sortedMap = new TreeMap<>();
+                            sortedMap.put(1, 2L);
+                            StringUtils.mostCommonIndent(sortedMap);
+                        }
+                    }
+            """.trimIndent(),
+            checkCompilationUnit = { compilationUnit ->
+                //Make sure the method invocation to "mostCommonIndent" has a type:
+                val method = compilationUnit.classes[0].body.statements[0] as J.MethodDeclaration
+                val methodInvocation = method.body?.statements?.get(2) as J.MethodInvocation
+                assertThat(methodInvocation.type).isNotNull
+            }
+        )
+    }
+
+    @Test
+    fun testMethodTypeParameters() {
+        assertCompilationWithByteCode(
+            parser = JavaParser.fromJavaVersion()
```

Comment:
Should we be following the TCK pattern we use elsewhere to run this against all of the JDKs we support, rather than whatever JDK we happen to use when running tests?

Reply:
I believe we are doing this, unless I have missed something?

Reply:
To do this, the parser would need to be an input to the method, the class should be an interface instead, and concrete implementations drawn up in the two places where all such interfaces are implemented.

I'd also like us to not use the redundant `test` prefix to test method names for consistency with other tests.